### PR TITLE
Irdl generic types coercions

### PR DESCRIPTION
**Pull request by math-fehr**
This PR will detach IRDL from the different coercion definitions for generic attribute parameters. This PR does not remove nor add any functionality.

Before this PR, each generic data structure used in a `Data` should define in IRDL a way to convert their type (like `ListAttr[T]`) to a corresponding constraint (like `ArrayOf(T)`).

With this PR, each `Data` definition can extend `IRDLGenericCoercion` to provide a translation from their type to a constraint. For instance, this is the corresponding definition of `ArrayAttr`:

```python
@irdl_attr_definition
class ArrayAttr(Data[List[A]], IRDLGenericCoercion):
    ...

    @staticmethod
    def generic_constraint_coercion(args: tuple[Any]) -> AttrConstraint:
        if len(args) == 0:
            return ArrayOfConstraint(irdl_to_attr_constraint(args[0]))
        if len(args) == 1:
            return ArrayOfConstraint(AnyAttr())
        raise TypeError(f"Attribute ArrayAttr expects at most type"
                        f" parameter, but {len(args)} were given")

```

### CODE REVIEW

File: src/xdsl/ir.py [hunk 1]
```diff
@@ -196,7 +196,7 @@ def build(cls: Type[A], *args: Any) -> A:
 
 
 @dataclass(frozen=True)
-class Data(Generic[DataElement], Attribute):
+class Data(Generic[DataElement], Attribute, ABC):
```
Comment (michel-steuwer) [inline src/xdsl/ir.py:199]: I don't get why this change is necessary.

**Review by michel-steuwer — COMMENTED** [review]
I think the if-conditions needs changing, otherwise looks ok.
More comments in the other PR.

File: src/xdsl/dialects/builtin.py [hunk 2]
```diff
@@ -170,31 +188,21 @@ def print_parameter(data: List[A], printer: Printer) -> None:
         printer.print_list(data, printer.print_attribute)
         printer.print_string("]")
 
+    @staticmethod
+    def generic_constraint_coercion(args: tuple[Any]) -> AttrConstraint:
+        if len(args) == 0:
+            return ArrayOfConstraint(irdl_to_attr_constraint(args[0]))
```
[the remark this answers is not in the document]
Comment (math-fehr) [reply src/xdsl/dialects/builtin.py:194]: Oh yeah, I inverted the two cases!
The `len(args) == 0` case correspond to `ArrayAttr[Attribute]`, it was just a shorthand. But we can remove it if you feel that this is confusing / unneeded!

File: src/xdsl/ir.py (hunk 1 above)
Comment (math-fehr) [reply src/xdsl/ir.py:199]: Oh, I just added it because I saw it missing. From what I understand, it improves error messages on IDE (and possibly at runtime).

**Comment by math-fehr** [comment]
Okay, this should look better in my opinion, tell me what you think!

**Comment by math-fehr** [comment]
@michel-steuwer, do you think I can merge it to the other PR?

**Review by michel-steuwer — APPROVED** [review]
